README.md
# blessed

A curses-like library with a high level terminal interface API for node.js.

![blessed](https://raw.githubusercontent.com/chjj/blessed/master/img/v0.1.0-3.gif)

Blessed is over 16,000 lines of code and terminal goodness. It's completely
implemented in javascript, and its goal consists of two things:

1. Reimplement ncurses entirely by parsing and compiling terminfo and termcap,
and exposing a `Program` object which can output escape sequences compatible
with _any_ terminal.

2. Implement a widget API which is heavily optimized for terminals.

The blessed renderer makes use of CSR (change-scroll-region), and BCE
(back-color-erase). It draws the screen using the painter's algorithm and is
sped up with smart cursor movements and a screen damage buffer. This means
rendering of your application will be extremely efficient: blessed only draws
the changes (damage) to the screen.

Blessed is arguably as accurate as ncurses, but even more optimized in some
ways. The widget library gives you an API which is reminiscent of the DOM.
Anyone is able to make an awesome terminal application with blessed. There are
terminal widget libraries for other platforms (primarily [python][urwid] and
[perl][curses-ui]), but blessed is possibly the most DOM-like (dare I say the
most user-friendly?).

Blessed has been used to implement other popular libraries and programs.
Examples include: the [slap text editor][slap] and [blessed-contrib][contrib].
The blessed API itself has gone on to inspire [termui][termui] for Go.

## Important Blessed Changes (>0.0.51)

- The absolute `.left` _property_ (not option) has been renamed to `.aleft`.
  The `.rleft` property has been renamed to `.left`. This should not have much
  effect on most applications. This includes all other coordinate properties.
- `autoPadding` is now enabled by default. To revert to the original behavior,
  pass `autoPadding: false` into the screen object. That being said, it would
  be wiser to adjust your code to use `autoPadding`. non-`autoPadding` is now
  considered deprecated.

## Install

``` bash
$ npm install blessed
```

## Example

This will render a box with line borders containing the text `'Hello world!'`,
perfectly centered horizontally and vertically.

__NOTE__: It is recommend you use either `smartCSR` or `fastCSR` as a
`blessed.screen` option. `autoPadding` is also recommended; it will
automatically offset box content within borders instead of on top of them when
coords are `0`. non-`autoPadding` _may_ be deprecated in the future. See the
API documentation for further explanation of these options.

``` js
var blessed = require('blessed');

// Create a screen object.
var screen = blessed.screen({
  autoPadding: true,
  smartCSR: true
});

screen.title = 'my window title';

// Create a box perfectly centered horizontally and vertically.
var box = blessed.box({
  top: 'center',
  left: 'center',
  width: '50%',
  height: '50%',
  content: 'Hello {bold}world{/bold}!',
  tags: true,
  border: {
    type: 'line'
  },
  style: {
    fg: 'white',
    bg: 'magenta',
    border: {
      fg: '#f0f0f0'
    },
    hover: {
      bg: 'green'
    }
  }
});

// Append our box to the screen.
screen.append(box);

// Add a PNG icon to the box (X11 only)
var icon = blessed.image({
  parent: box,
  top: 0,
  left: 0,
  width: 'shrink',
  height: 'shrink',
  file: __dirname + '/my-program-icon.png',
  search: false
});

// If our box is clicked, change the content.
box.on('click', function(data) {
  box.setContent('{center}Some different {red-fg}content{/red-fg}.{/center}');
  screen.render();
});

// If box is focused, handle `enter`/`return` and give us some more content.
box.key('enter', function(ch, key) {
  box.setContent('{right}Even different {black-fg}content{/black-fg}.{/right}\n');
  box.setLine(1, 'bar');
  box.insertLine(1, 'foo');
  screen.render();
});

// Quit on Escape, q, or Control-C.
screen.key(['escape', 'q', 'C-c'], function(ch, key) {
  return process.exit(0);
});

// Focus our element.
box.focus();

// Render the screen.
screen.render();
```


## API Documentation

### Widgets

Blessed comes with a number of high-level widgets so you can avoid all the
nasty low-level terminal stuff.

- [Node](#node-from-eventemitter)
- [Screen](#screen-from-node)
- [Element](#element-from-node)
- [Box](#box-from-element)
- [Text](#text-from-element)
- [Line](#line-from-box)
- [ScrollableBox](#scrollablebox-from-box)
- [ScrollableText](#scrollabletext-from-scrollablebox)
- [List](#list-from-box)
- [Form](#form-from-box)
- [Input](#input-from-box)
- [Textarea](#textarea-from-input)
- [Textbox](#textbox-from-textarea)
- [Button](#button-from-input)
- [ProgressBar](#progressbar-from-input)
- [FileManager](#filemanager-from-list)
- [Checkbox](#checkbox-from-input)
- [RadioSet](#radioset-from-box)
- [RadioButton](#radiobutton-from-checkbox)
- [Prompt](#prompt-from-box)
- [Question](#question-from-box)
- [Message](#message-from-box)
- [Loading](#loading-from-box)
- [Listbar](#listbar-from-box)
- [Log](#log-from-scrollabletext)
- [Table](#table-from-box)
- [ListTable](#listtable-from-list)
- [Terminal](#terminal-from-box)
- [Image](#image-from-box)


#### Node (from EventEmitter)

The base node which everything inherits from.

##### Options:

- __screen__ - the screen to be associated with.
- __parent__ - the desired parent.
- __children__ - an arrray of children.

##### Properties:

- inherits all from EventEmitter.
- __type__ - type of the node (e.g. `box`).
- __options__ - original options object.
- __parent__ - parent node.
- __screen__ - parent screen.
- __children__ - array of node's children.
- __data, _, $__ - an object for any miscellanous user data.
- __index__ - render index (document order index) of the last render call.

##### Events:

- inherits all from EventEmitter.
- __adopt__ - received when node is added to a parent.
- __remove__ - received when node is removed from it's current parent.
- __reparent__ - received when node gains a new parent.
- __attach__ - received when node is attached to the screen directly or
  somewhere in its ancestry.
- __detach__ - received when node is detached from the screen directly or
  somewhere in its ancestry.

##### Methods:

- inherits all from EventEmitter.
- __prepend(node)__ - prepend a node to this node's children.
- __append(node)__ - append a node to this node's children.
- __remove(node)__ - remove child node from node.
- __insert(node, i)__ - insert a node to this node's children at index `i`.
- __insertBefore(node, refNode)__ - insert a node to this node's children
  before the reference node.
- __insertAfter(node, refNode)__ - insert a node from node after the reference
  node.
- __detach()__ - remove node from its parent.
- __emitDescendants(type, args..., [iterator])__ - emit event for element, and
  recursively emit same event for all descendants.
- __get(name, [default])__ - get user property with a potential default value.
- __set(name, value)__ - set user property to value.


#### Screen (from Node)

The screen on which every other node renders.

##### Options:

- __program__ - the blessed `Program` to be associated with. will be
  automatically instantiated if none is provided.
- __smartCSR__ - attempt to perform CSR optimization on all possible elements
  (not just full-width ones, elements with uniform cells to their sides).
  this is known to cause flickering with elements that are not full-width,
  however, it is more optimal for terminal rendering.
- __fastCSR__ - do CSR on any element within 20 cols of the screen edge on
  either side. faster than `smartCSR`, but may cause flickering depending on
  what is on each side of the element.
- __useBCE__ - attempt to perform `back_color_erase` optimizations for terminals
  that support it. it will also work with terminals that don't support it, but
  only on lines with the default background color. as it stands with the current
  implementation, it's uncertain how much terminal performance this adds at the
  cost of overhead within node.
- __resizeTimeout__ - amount of time (in ms) to redraw the screen after the
  terminal is resized (default: 300).
- __tabSize__ - the width of tabs within an element's content.
- __autoPadding__ - automatically position child elements with border and
  padding in mind (__NOTE__: this is a recommended option. it may become
  default in the future).
- __artificialCursor__ - have blessed draw a custom cursor and hide the
  terminal cursor (__experimental__).
- __cursorShape__ - shape of the artificial cursor. can be: block, underline,
  or line.
- __cursorBlink__ - whether the artificial cursor blinks.
- __cursorColor__ - color of the artificial color. accepts any valid color
  value (`null` is default).
- __log__ - create a log file. see `log` method.
- __dump__ - dump all output and input to desired file. can be used together
  with `log` option if set as a boolean.
- __debug__ - debug mode. enables usage of the `debug` method. also creates a
  debug console which will display when pressing F12. it will display all log
  and debug messages.
- __ignoreLocked__ - Array of keys in their full format (e.g. `C-c`) to ignore
  when keys are locked. Useful for creating a key that will _always_ exit no
  matter whether the keys are locked.
- __dockBorders__ - automatically "dock" borders with other elements instead of
  overlapping, depending on position (__experimental__). for example:
  These border-overlapped elements:
```
┌─────────┌─────────┐
│ box1    │ box2    │
└─────────└─────────┘
```
  Become:
```
┌─────────┬─────────┐
│ box1    │ box2    │
└─────────┴─────────┘
```
- __fullUnicode__ - allow for rendering of East Asian double-width characters,
  utf-16 surrogate pairs, and unicode combining characters. this allows you to
  display text above the basic multilingual plane. this is behind an option
  because it may affect performance slightly negatively. without this option
  enabled, all double-width, surrogate pair, and combining characters will be
  replaced by `??`, `?`, `` respectively. (NOTE: libvte (e.g. gnome-terminal)
  cannot display characters that are both surrogate pairs _and_ double-width
  properly. there is no way for blessed to fix this unfortunately).

##### Properties:

- inherits all from Node.
- __program__ - the blessed Program object.
- __tput__ - the blessed Tput object (only available if you passed `tput: true`
  to the Program constructor.)
- __focused__ - top of the focus history stack.
- __width__ - width of the screen (same as `program.cols`).
- __height__ - height of the screen (same as `program.rows`).
- __cols__ - same as `screen.width`.
- __rows__ - same as `screen.height`.
- __left__ - relative left offset, always zero.
- __right__ - relative right offset, always zero.
- __top__ - relative top offset, always zero.
- __bottom__ - relative bottom offset, always zero.
- __aleft__ - absolute left offset, always zero.
- __aright__ - absolute right offset, always zero.
- __atop__ - absolute top offset, always zero.
- __abottom__ - absolute bottom offset, always zero.
- __grabKeys__ - whether the focused element grabs all keypresses.
- __lockKeys__ - prevent keypresses from being received by any element.
- __hover__ - the currently hovered element. only set if mouse events are bound.
- __title__ - set or get window title.

##### Events:

- inherits all from Node.
- __resize__ - received on screen resize.
- __mouse__ - received on mouse events.
- __keypress__ - received on key events.
- __element [name]__ - global events received for all elements.
- __key [name]__ - received on key event for [name].
- __focus, blur__ - received when the terminal window focuses/blurs. requires a
  terminal supporting the focus protocol and focus needs to be passed to
  program.enableMouse().
- __prerender__ - received before render.
- __render__ - received on render.

##### Methods:

- inherits all from Node.
- __log(msg, ...)__ - write string to the log file if one was created.
- __debug(msg, ...)__ - same as the log method, but only gets called if the
  `debug` option was set.
- __alloc()__ - allocate a new pending screen buffer and a new output screen
  buffer.

CHANGELOG.md
# Blessed v0.1.0 - new terminal goodies for node.js

![blessed](https://raw.githubusercontent.com/chjj/blessed/master/img/v0.1.0-3.gif)

The features demonstrated in the above gif __element transparency/shadow__ and
__border docking__.

## New useful options for your typewriter application:

- __`transparent` option__ - Lower element opacity to 50%. This will display
  dimmed elements and content behind the foreground element using a naive color
  blending function (good enough for a terminal's limited amount of colors).
  works best with 256color terminals. (see widget-shadow.js)

- __`shadow` option__ - Give the element a translucent shadow. Automatically
  darkens the background behind it. (see widget-shadow.js)

- __`dockBorders` option__ - Element borders will automatically "dock" to each
  other. Instead of overlapping the borders end up connecting. (see
  widget-dock.js)

- __`autoPadding` default__ - Auto padding is now enabled by default, meaning
  blessed will automatically position elements inside their parent's border.

- __`rleft` property__ - Relative offsets are now default element properties
  (`left` instead of `rleft`).

- __`draggable` property__ - Make any element draggable with the mouse. (see
  widget-shadow.js or widget-dock.js)

- __`Table` and `ListTable` elements__ - Tables with a high quality rendering.
  (see widget-table.js and widget-listtable.js)

- __`Log` element__ - A top to bottom logger box with scrollback and other
  features. (see widget-log.js)

- __Obscurable borders__ - In addition to docking borders, it's possible to
  obscure borders by sliding them off the screen with negative offsets. (see
  widget-dock.js)

- __Percentage expressions__ - Like CSS, arithmetic can now be performed on
  percentages. e.g. `width: '50%-1'`. This is useful for overlapping borders on
  elements with a percentage width. (see widget-dock.js)

## Other features that weren't mentioned before:

- __`setHover` option__ - Set a hover text box to follow cursor on mouseover,
  similar to how a web browser handles the "title" attribute. (see widget.js)

- __`Terminal` element__ - Spin up a pseudo terminal as a blessed element.
  useful for writing a terminal multiplexer. (requires term.js and pty.js as
  optional dependencies). (see example/multiplex.js)

- __`Image` element__ - Uses `w3mimgdisplay` to draw real images your terminal.
  this is much easier than calling w3mimgdisplay by hand. Image elements behave
  like any other element, although it is wise to use `width: 'shrink', height:
  'shrink'`. (see widget-image.js)

---

The major things that justified the 0.1.0 release were fixes and stabilization
of api (`autoPadding`/`rleft`/`left`). Scrolling boxes were almost completely
revamped to work a bit smarter.

---

## Things yet to come:

- __@secrettriangle's [improvements](https://github.com/slap-editor/slap) for
  textareas__ - This allows for real text navigation.

- __Gravity and margin layouts__

This is something that's been in the idea bin for a while. Every element could
potentially have properties like:

```
  gravity: 'bottomleft',
  margin: 5,
``

In other words, just a more complex `float` system than what the CSSOM is used
to.


lib/high.js
/**
 * Blessed high-level interface
 * Still in development
 */

/*
  API Example:
  This will render a box with ascii borders containing the
  text 'Hello world!', centered horizontally and vertically.

  var blessed = require('blessed')
    , program = blessed()
    , screen;

  screen = new blessed.Screen({
    program: program
  });

  screen.append(new blessed.Text({
    screen: screen,
    parent: screen,
    fg: 3,
    bg: 5,
    border: {
      type: 'ascii',
      fg: 1
    },
    content: 'Hello world!',
    top: 'center',
    left: 'center'
  }));

  program.on('keypress', function(ch, key) {
    if (key.name === 'escape') {
      process.exit(0);
    }
  });

  screen.render();
*/

/**
 * Node
 */

function Node(options) {
  this.children = options.children || [];
}

Node.prototype.prepend = function(element) {
  this.children.unshift(element);
};

Node.prototype.append = function(element) {
  this.children.push(element);
};

Node.prototype.remove = function(element) {
  var i = this.children.indexOf(element);
  if (~i) this.children.splice(i, 1);
};

Node.prototype.detach = function(element) {
  this.parent.remove(element);
};

/**
 * Screen
 */

function Screen(options) {
  var self = this;

  Node.call(this, options);

  this.children = options.children || [];
  this.program = options.program;
  this.tput = this.program.tput;
  this.dattr = ((0 << 18) | (0x1ff << 9)) | 0x1ff;
  this.position = {
    left: this.left = 0,
    right: this.right = 0,
    top: this.top = 0,
    bottom: this.bottom = 0
  };

  this.alloc();

  this.program.on('resize', function() {
    self.alloc();
    self.render();
  });
}

Screen.prototype.__proto__ = Node.prototype;

Screen.prototype.__defineGetter__('cols', function() {
  return this.program.cols;
});

Screen.prototype.__defineGetter__('rows', function() {
  return this.program.rows;
});

Screen.prototype.__defineGetter__('width', function() {
  return this.program.cols;
});

Screen.prototype.__defineGetter__('height', function() {
  return this.program.rows;
});

Screen.prototype.alloc = function() {
  var x, y;
  this.lines = [];
  for (y = 0; y < this.rows; y++) {
    this.lines[y] = [];
    for (x = 0; x < this.cols; x++) {
      this.lines[y][x] = [this.dattr, ' '];
    }
    this.lines[y].dirty = true;
  }
};

Screen.prototype.render = function() {
  this.children.forEach(function(el) {
    el.render();
  });
  this.draw(0, this.rows - 1);
};

Screen.prototype.draw = function(start, end) {
  var x
    , y
    , line
    , out
    , ch
    , data
    , attr
    , fgColor
    , bgColor
    , flags;

  this.program.saveCursor();

  for (y = start; y <= end; y++) {
    line = this.lines[y];

    if (!line.dirty) continue;
    line.dirty = false;

    out = '';
    attr = this.dattr;

    for (x = 0; x < this.cols; x++) {
      data = line[x][0];
      ch = line[x][1];

      if (data && data !== attr) {
        if (attr !== this.dattr) {
          out += '\x1b[m';
        }
        if (data !== this.dattr) {
          out += '\x1b[';

          bgColor = data & 0x1ff;
          fgColor = (data >> 9) & 0x1ff;
          flags = data >> 18;

          // bold
          if (flags & 1) {
            out += '1;';
          }

          // underline
          if (flags & 2) {
            out += '4;';
          }

          // blink
          if (flags & 4) {
            out += '5;';
          }

          // inverse
          if (flags & 8) {
            out += '7;';
          }

          // invisible
          if (flags & 16) {
            out += '8;';
          }

          if (bgColor !== 0x1ff) {
            if (bgColor < 16) { //|| this.tput.colors <= 16) {
              if (bgColor < 8) {
                bgColor += 40;
              } else if (bgColor < 16) {
                bgColor += 100;
              }
              out += bgColor + ';';
            } else {
              out += '48;5;' + bgColor + ';';
            }
          }

          if (fgColor !== 0x1ff) {
            if (fgColor < 16) { //|| this.tput.colors <= 16) {
              if (fgColor < 8) {
                fgColor += 30;
              } else if (fgColor < 16) {
                fgColor += 90;
              }
              out += fgColor + ';';
            } else {
              out += '38;5;' + fgColor + ';';
            }
          }

          if (out[out.length-1] === ';') out = out.slice(0, -1);

          out += 'm';
        }
      }

      out += ch;
      attr = data;
    }

    if (attr && attr !== this.dattr) {
      out += '\x1b[m';
    }

    //this.program.move(y + 1, 1);
    //this.program.write(out);
    this.program.write('\x1b[' + (y + 1) + ';1H' + out);
  }

  this.program.restoreCursor();
};

/**
 * Element
 */

function Element(options) {
  options = options || {};

  Node.call(this, options);

  this.options = options;

  this.screen = options.screen;
  this.parent = options.parent || (function(){throw Error('No parent.')})();
  this.position = {
    left: options.left || 0,
    right: options.right || 0,
    top: options.top || 0,
    bottom: options.bottom || 0,
    width: options.width || null,
    height: options.height || null
  };

  this.fg = options.fg || 0x1ff;
  this.bg = options.bg || 0x1ff;
  this.bold = options.bold ? 1 : 0;
  this.underline = options.underline ? 2 : 0;

  if (this.fg === -1) this.fg = exports.NORMAL;
  if (this.bg === -1) this.bg = exports.NORMAL;

  this.fixed = options.fixed || false;
  this.border = options.border;
  if (this.border) {
    this.border.type = this.border.type || 'bg';
    this.border.fg = this.border.fg || -1;
    this.border.bg = this.border.bg || -1;
    this.border.ch = this.border.ch || ' ';
    this.border.bold = this.border.bold ? 1 : 0;
    this.border.underline = this.border.underline ? 2 : 0;
    if (this.border.fg === -1) this.border.fg = exports.NORMAL;
    if (this.border.bg === -1) this.border.bg = exports.NORMAL;
  }

  this.children = options.children || [];

  this.clickable = this.clickable || false;
  this.hover = this.hover || false;

  this.content = options.content || '';
}

Element.prototype.__proto__ = Node.prototype;

Element.prototype.__defineGetter__('left', function() {
  var left = this.position.left;

  if (typeof left === 'string') {
    if (left === 'center') left = '50%';
    left = +left.slice(0, -1) / 100;
    var len = Math.min(this.width, this.parent.width);
    left = (this.parent.width - len) * left | 0;
  }

  var opt = this.options;
  if (opt.left == null && !left && opt.right != null && this.position.right) {
    left = this.width - this.position.right;
  }

  return (this.parent.left || 0) + left;
});

Element.prototype.__defineSetter__('left', function(val) {
  return this.position.left = val;
});

Element.prototype.__defineGetter__('right', function() {
  return (this.parent.right || 0) + this.position.right;
});

Element.prototype.__defineSetter__('right', function(val) {
  return this.position.right = val;
});

Element.prototype.__defineGetter__('top', function() {
  var top = this.position.top;

  if (typeof top === 'string') {
    if (top === 'center') top = '50%';
    top = +top.slice(0, -1) / 100;
    var len = Math.min(this.height, this.parent.height);
    top = (this.parent.height - len) * top | 0;
  }

  var opt = this.options;
  if (opt.top == null && !top && opt.bottom != null && this.position.bottom) {
    top = this.height - this.position.bottom;
  }

  if (this.parent.childBase != null && ~this.parent.items.indexOf(this)) {
    top += this.parent.items.indexOf(this);
    top -= this.parent.childBase;
  }

  return (this.parent.top || 0) + top;
});

Element.prototype.__defineSetter__('top', function() {
  return this.position.top = val;
});

Element.prototype.__defineGetter__('bottom', function() {
  return (this.parent.bottom || 0) + this.position.bottom;
});

Element.prototype.__defineSetter__('bottom', function(val) {
  return this.position.bottom = val;
});

Element.prototype.__defineGetter__('width', function() {
  var width = this.position.width;
  if (typeof width === 'string') {
    if (width === 'half') width = '50%';
    width = +width.slice(0, -1) / 100;
    return this.parent.width * width | 0;
  }
  if (!width) {
    // Problem if .left is 'center', we can't calculate the width
    var left = this.position.left;
    if (typeof left === 'string') {
      if (left === 'center') left = '50%';
      left = +left.slice(0, -1) / 100;
      left = this.parent.width * left | 0;
    }
    width = this.parent.width - this.position.right - left;
  }
  return width;
});

Element.prototype.__defineSetter__('width', function(val) {
  return this.position.width = val;
});

Element.prototype.__defineGetter__('height', function() {
  var height = this.position.height;
  if (typeof height === 'string') {
    if (height === 'half') height = '50%';
    height = +height.slice(0, -1) / 100;
    return this.parent.height * height | 0;
  }
  if (!height) {
    // Problem if .top is 'center', we can't calculate the height
    var top = this.position.top;
    if (typeof top === 'string') {
      if (top === 'center') top = '50%';
      top = +top.slice(0, -1) / 100;
      top = this.parent.height * top | 0;
    }
    height = this.parent.height - this.position.bottom - top;
  }
  return height;
});

Element.prototype.__defineSetter__('height', function(val) {
  return this.position.height = val;
});

Element.prototype.__defineGetter__('rleft', function() {
  var left = this.position.left;

  if (typeof left === 'string') {
    if (left === 'center') left = '50%';
    left = +left.slice(0, -1) / 100;
    var len = Math.min(this.width, this.parent.width);
    left = len * left | 0;
  }

  var opt = this.options;
  if (opt.left == null && !left && opt.right != null && this.position.right) {
    left = this.width - this.position.right;
  }

  return left;
});

Element.prototype.__defineSetter__('rleft', function(val) {
  return this.position.left = val;
});

Element.prototype.__defineGetter__('rright', function() {
  return this.position.right;
});

Element.prototype.__defineSetter__('rright', function() {
  return this.position.right = val;
});

Element.prototype.__defineGetter__('rtop', function() {
  var top = this.position.top;

  if (typeof top === 'string') {
    if (top === 'center') top = '50%';
    top = +top.slice(0, -1) / 100;
    var len = Math.min(this.height, this.parent.height);
    top = len * top | 0;
  }

  var opt = this.options;
  if (opt.top == null && !top && opt.bottom != null && this.position.bottom) {
    top = this.height - this.position.bottom;
  }

  if (this.parent.childBase != null && ~this.parent.items.indexOf(this)) {
    top += this.parent.items.indexOf(this);
    top -= this.parent.childBase;
  }

  return top;
});

Element.prototype.__defineSetter__('rtop', function() {
  return this.position.top = val;
});

Element.prototype.__defineGetter__('rbottom', function() {
  return this.position.bottom;
});

Element.prototype.__defineSetter__('rbottom', function() {
  return this.position.bottom = val;
});

/**
 * Border Measures
 */

Element.prototype.__defineGetter__('bheight', function() {
  return this.height - (this.parent.border ? 1 : 0);
});

Element.prototype.__defineGetter__('bwidth', function() {
  return this.width - (this.parent.border ? 1 : 0);
});

Element.prototype.__defineGetter__('bleft', function() {
  return this.left + (this.parent.border ? 1 : 0);
});

Element.prototype.__defineGetter__('bright', function() {
  return this.right + (this.parent.border ? 1 : 0);
});

Element.prototype.__defineGetter__('btop', function() {
  return this.top + (this.parent.border ? 1 : 0);
});

Element.prototype.__defineGetter__('bbottom', function() {
  return this.bottom + (this.parent.border ? 1 : 0);
});

Element.prototype.__defineGetter__('brleft', function() {
  return this.rleft + (this.parent.border ? 1 : 0);
});

Element.prototype.__defineGetter__('brright', function() {
  return this.rright + (this.parent.border ? 1 : 0);
});

Element.prototype.__defineGetter__('brtop', function() {
  return this.rtop + (this.parent.border ? 1 : 0);
});

Element.prototype.__defineGetter__('brbottom', function() {
  return this.rbottom + (this.parent.border ? 1 : 0);
});

/**
 * Box
 */

function Box(options) {
  Element.call(this, options);
}

Box.prototype.__proto__ = Element.prototype;

Box.prototype.render = function() {
  var lines = this.screen.lines
    , xi = this.left
    , xl = this.screen.cols - this.right
    , yi = this.top
    , yl = this.screen.rows - this.bottom
    , cell
    , attr
    , ch
    , ci = 0
    , cl = this.content.length;

  if (this.position.width) {
    xl = xi + this.width;
  }

  if (this.position.height) {
    yl = yi + this.height;
  }

  for (yi = this.top; yi < yl; yi++) {
    for (xi = this.left; xi < xl; xi++) {
      if (this.border && (yi === this.top || xi === this.left || yi === yl - 1 || xi === xl - 1)) {
        attr = ((this.border.bold << 18) + (this.border.underline << 18)) | (this.border.fg << 9) | this.border.bg;
        if (this.border.type === 'ascii') {
          if (yi === this.top) {
            if (xi === this.left) ch = '┌';
            else if (xi === xl - 1) ch = '┐';
            else ch = '─';
          } else if (yi === yl - 1) {
            if (xi === this.left) ch = '└';
            else if (xi === xl - 1) ch = '┘';
            else ch = '─';
          } else if (xi === this.left || xi === xl - 1) {
            ch = '│';
          }
        } else if (this.border.type === 'bg') {
          ch = this.border.ch;
        }
      } else {
        attr = ((this.bold << 18) + (this.underline << 18)) | (this.fg << 9) | this.bg;
        ch = this.content[ci++] || ' ';
      }
      cell = lines[yi][xi];
      if (attr !== cell[0] || ch !== cell[1]) {
        lines[yi][xi][0] = attr;
        lines[yi][xi][1] = ch;
        lines[yi].dirty = true;
      }
    }
  }

  this.children.forEach(function(el) {
    el.render();
  });

  //var o = { x: 0, y: 0 };
  //this.children.forEach(function(el) {
  //  el.render(o);
  //});
};

/**
 * Text
 */

function Text(options) {
  Element.call(this, options);
  this.full = options.full;
  this.noOverflow = options.noOverflow;
}

Text.prototype.__proto__ = Element.prototype;

Text.prototype.render = function() {
  var lines = this.screen.lines
    , xi = this.left
    , xl = this.screen.cols - this.right
    , yi = this.top
    , yl = this.screen.rows - this.bottom
    , cell
    , attr
    , ch
    , ci = 0
    , cl = this.content.length
    , ended = -1;

  if (this.position.width) {
    xl = xi + this.width;
  }

  if (this.position.height) {
    yl = yi + this.height;
  }

  if (this.full) {
    xl = xi + this.parent.width - (this.parent.border ? 2 : 0) - this.rleft - this.rright;
  }

  if (this.noOverflow) {
    var rtop = this.rtop - (this.parent.border ? 1 : 0)
      , visible = this.parent.height - (this.parent.border ? 2 : 0);

    if (rtop < 0) {
      return;
    }

    if (rtop >= visible) {
      return;
    }
  }

  for (; yi < yl; yi++) {
    for (xi = this.left; xi < xl; xi++) {
      cell = lines[yi][xi];
      attr = ((this.bold << 18) + (this.underline << 18)) | (this.fg << 9) | this.bg;
      ch = this.content[ci++];
      if (!ch) {
        if (this.full) {
          if (ended === -1) ended = yi;
          if (yi === ended) {
            ch = ' ';
          } else {
            break;
          }
        } else {
          break;
        }
      }
      if (attr !== cell[0] || ch !== cell[1]) {
        lines[yi][xi][0] = attr;
        lines[yi][xi][1] = ch;
        lines[yi].dirty = true;
      }
    }
  }
};

/**
 * ScrollableBox
 */

function ScrollableBox(options) {
  Box.call(this, options);
  this.childOffset = 0;
  this.childBase = 0;
}

ScrollableBox.prototype.__proto__ = Box.prototype;

/**
 * List
 */

function List(options) {
  ScrollableBox.call(this, options);

  this.items = [];
  this.selected = 0;

  this.selectedBg = options.selectedBg || -1;
  this.selectedFg = options.selectedFg || -1;
  this.selectedBold = options.selectedBold ? 1 : 0;
  this.selectedUnderline = options.selectedUnderline ? 2 : 0;

  if (this.selectedBg === -1) this.selectedBg = exports.NORMAL;
  if (this.selectedFg === -1) this.selectedFg = exports.NORMAL;

  if (options.items) {
    options.items.forEach(this.add.bind(this));
  }

  if (this.children.length) {
    this.select(0);
  }
}

List.prototype.__proto__ = ScrollableBox.prototype;

List.prototype.add = function(item) {
  var item = new Text({
    screen: this.screen,
    parent: this,
    fg: this.fg,
    bg: this.bg,
    content: item.content || item,
    top: 0,
    left: (this.border ? 1 : 0) + 1,
    full: true,
    noOverflow: true,
    height: 1
  });
  this.append(item);
  this.items.push(item);
};

List.prototype._remove = List.prototype.remove;
List.prototype.remove = function(child) {
  if (typeof child === 'number') {
    child = this.children[child];
  }
  var i = this.items.indexOf(child);
  if (~i) this.items.splice(i, 1);
  this._remove(child);
};

List.prototype.select = function(index) {
  if (!this.items.length) return;

  if (typeof index === 'object') {
    index = this.items.indexOf(index);
  }

  if (index < 0) index = 0;
  else if (index >= this.items.length) index = this.items.length - 1;

  if (this.selected === index && this._listInitialized) return;
  this._listInitialized = true;

  if (this.selectedBg) {
    this.items[this.selected].bg = this.bg;
    this.items[index].bg = this.selectedBg;
  }

  if (this.selectedFg) {
    this.items[this.selected].fg = this.fg;
    this.items[index].fg = this.selectedFg;
  }

  if (this.selectedBold != null) {
    this.items[this.selected].bold = this.bold;
    this.items[index].bold = this.selectedBold;
  }

  if (this.selectedUnderline != null) {
    this.items[this.selected].underline = this.underline;
    this.items[index].underline = this.selectedUnderline;
  }

  var diff = index - this.selected;
  this.childOffset += diff;

  this.selected = index;

  var visible = this.height - (this.border ? 2 : 0);
  if (this.childOffset > visible - 1) {
    var d = this.childOffset - (visible - 1);
    this.childOffset -= d;
    this.childBase += d;
  } else if (this.childOffset < 0) {
    var d = this.childOffset;
    this.childOffset += -d;
    this.childBase += d;
  }
};

List.prototype.move = function(offset) {
  this.select(this.selected + offset);
};

List.prototype.up = function(offset) {
  this.move(-(offset || 1));
};

List.prototype.down = function(offset) {
  this.move(offset || 1);
};

/**
 * Input
 */

function Input(options) {
  Box.call(this, options);
}

Input.prototype.__proto__ = Box.prototype;

/**
 * Textbox
 */

function Textbox(options) {
  Input.call(this, options);
}

Textbox.prototype.__proto__ = Input.prototype;

/**
 * Button
 */

function Button(options) {
  Input.call(this, options);
}

Button.prototype.__proto__ = Input.prototype;

/**
 * ProgressBar
 */

function ProgressBar(options) {
  Input.call(this, options);
  this.filled = options.filled || 0;
  if (typeof this.filled === 'string') {
    this.filled = +this.filled.slice(0, -1);
  }
  this.ch = options.ch || ' ';
  this.barFg = options.barFg || -1;
  this.barBg = options.barBg || -1;
  if (this.barFg === -1) this.barFg = exports.NORMAL;
  if (this.barBg === -1) this.barBg = exports.NORMAL;
}

ProgressBar.prototype.__proto__ = Input.prototype;

ProgressBar.prototype._render = ProgressBar.prototype.render;
ProgressBar.prototype.render = function() {
  this._render();

  var x, y;

  var h = this.height
    , w = this.width * (this.filled / 100)
    , l = this.left
    , t = this.top;

  for (y = t; y < t + h; y++) {
    for (x = l; x < l + w; x++) {
      attr = ((this.bold << 18) + (this.underline << 18)) | (this.barFg << 9) | this.barBg;
      ch = this.ch;
      cell = lines[yi][xi];
      if (attr !== cell[0] || ch !== cell[1]) {
        lines[yi][xi][0] = attr;
        lines[yi][xi][1] = ch;
        lines[yi].dirty = true;
      }
    }
  }
};

ProgressBar.prototype.progress = function(filled) {
  this.filled += filled;
  if (this.filled < 0) this.filled = 0;
  else if (this.filled > 100) this.filled = 100;
};

/**
 * Constants
 */

exports.NORMAL = 0x1ff;

/**
 * Expose
 */

exports.Screen = Screen;
exports.Box = Box;
exports.Text = Text;
exports.ScrollableBox = ScrollableBox;
exports.List = List;
exports.Input = Input;
exports.Textbox = Textbox;
exports.Button = Button;
exports.ProgressBar = ProgressBar;
